src/rust/ide/src/executor/delay.rs

```diff
@@ -0,0 +1,82 @@
+//! Module defining `FrameDelay` - a `Future` that wraps a closure that should be called at a later
+//! time.
+
+pub use enso_prelude::*;
+
+use enso_protocol::prelude::Future;
+use ensogl_system_web::prelude::Closure;
+use ensogl_system_web::request_animation_frame;
+use futures::task::Context;
+use futures::task::Poll;
+use futures::task::Waker;
+use std::pin::Pin;
+
+
+// ========================
+// === Type Definitions ===
+// ========================
+
+type OnFrameClosure = Closure<dyn FnMut(f64)>;
+
+
+
+// ===================
+// === Frame Delay ===
+// ===================
+
+
+/// Delayed calls to a closure. The time delays is specified in number of render frames.
+#[derive(Debug)]
+pub struct FrameDelay<T>{
+    /// Counter that keeps track of the passage of frames. Needs to be shared with the JS runtime.
+    frames_left : Rc<RefCell<u64>>,
+    /// Internal closure that is shared with the JS runtime and updates the `frames_left`.
+    on_frame    : RefCell<Option<OnFrameClosure>>,
+    /// Target callback that is called after the delay has passed.
+    target      : T,
+}
+
+impl<T> FrameDelay<T> {
+    /// Constructor. Takes the number of render frames and the closure that should be called after
+    /// the render frames have been called.
+    pub fn new(n:u64, target:T) -> Self {
+        let frames_left = Rc::new(RefCell::new(n));
+        let on_frame    = default();
+        FrameDelay{frames_left,target,on_frame}
+    }
+
+    fn on_frame_action_init_if_uninit(&self, waker: &Waker) {
+        if self.on_frame.borrow().is_none() {
+            let waker       = waker.clone();
+            let frames_left = Rc::clone(&self.frames_left);
+            let closure     = Closure::new(move |_:f64| {
+                let decreased = frames_left.borrow().saturating_sub(1);
+                frames_left.replace(decreased);
+                waker.clone().wake();
```

Comment:
I'd like to have waker actually called when frame_left is equal to 0, otherwise this closure could reregister itself by request_animation_frame. Then we relieve executor from rescheduling this future on each frame.

Reply:
After discussion on PM we decided to leave it as it is - the complexity of my proposal is not adequate to the profit.